### ISSUE TITLE
WP-CLI + default WordPress email address

### ISSUE THREAD

**Issue by szepeviktor** [benhuson/wp-mailfrom#8]
Would you test your plugin when wp-cron is executed from cron/shell (not by a HTTP request)?
And also with wp-cli runs.

**Comment by szepeviktor** [comment]
In case of an invalid sender address I do
```php
        // Correct invalid From: address
        if ( null !== $this->from && ! $mail->validateAddress( $this->from ) ) {
            $mail->From = get_bloginfo( 'admin_email' );
        }
```
https://github.com/szepeviktor/smtp-uri/blob/master/smtp-uri.php#L73-L76

**Comment by benhuson** [comment]
The WordPress fix has been milestoned for WordPress 4.4
https://core.trac.wordpress.org/ticket/25239

Please see my fix for the plugin based on that core patch.
It continues to  use the existing WordPress method for creating the default from address based on $_SERVER['SERVER_NAME'], but if the result of that is not a valid email address then it used the site URL as per the core patch. 
https://github.com/benhuson/wp-mailfrom/commit/03a1f8a125748a7ecd5550cc2dfd27f99b88a228

Basically we need to ensure that whatever we thinking the default from email is is the same as what core thinks it is.

Please let me know what you think of my patch above.

Thanks

**Comment by szepeviktor** [comment]
I see that core developers think it is a good idea to set from to `wordpress@` + domain.
I do website maintenance - and even though I use `-all` SPF - I set `admin_email` as default from address in the SMPT URI plugin.

**Comment by szepeviktor** [comment]
But all together your plugin is not for defaults.

**Comment by benhuson** [comment]
Yep, the aim of this plugin it just to allow you to replace the default email address that WordPress specifies with something else.

So basically whatever WordPress sets as the default, we need to detect that so we can replace it.

The plugin does't allow you to set a default at present - we'd need to do that reliably without affecting where a different from address may already be specified.

PS. And I'd rather do everything via available filters. I don't plan on replacing the pluggable function for such trivial functionality - it would break other plugins that do replace the pluggable function for sending via SMTP etc.

**Comment by benhuson** [comment]
@szepeviktor I have added a 'wp_mailfrom_ii_default_from' filter which allows you to alter the default Wordpress email address that the plugin is expecting to change.
https://github.com/benhuson/wp-mailfrom/commit/ca2b875f20aaa801dd047696ba3f69fa1800ad14

As wp_mail() is a pluggable function, it's possible a replacement function might use a different default email address - do you do that in your plugin? If so you can use this 'wp_mailfrom_ii_default_from' filter to mirror the default email address you use in wp_mail().

Just to reiterate, the 'wp_mailfrom_ii_default_from' email address should always match the email address used in wp_mail(). So if WordPress created an invalid default email address when using WP-CLI, then this plugin should expect that invalid email address too.

I'll leave this ticket open so that when the default email address is changed in WordPress core (currently under review), I am reminded to reflect the plugin to change this:
https://core.trac.wordpress.org/ticket/25239

I plan to push out this filter as version 1.1 soon which should give you the flexibility if you need it for your plugin.

**Comment by szepeviktor** [comment]
Thank you very much.

1. This is a core bug
1. This is not related to my plugin
1. An invalid address cannot be tolerated

As I've noted above I only care about from address validity.
```php
        // Correct invalid From: address
        if ( null !== $this->from && ! $mail->validateAddress( $this->from ) ) {
            $mail->From = get_bloginfo( 'admin_email' );
        }
```
So whether it is a default address, empty address, invalid address
it should be corrected.

I hope 4.4 won't generate invalid from address.

**Comment by benhuson** [comment]
Doesn't look like this will be fixed in WordPress 4.4 :disappointed: 
https://core.trac.wordpress.org/ticket/25239#comment:17

**Comment by benhuson** [comment]
I notice this change has been made in core - in WP 5.5 I think:
https://core.trac.wordpress.org/ticket/25239#comment:114
https://core.trac.wordpress.org/changeset/48601

I guess that means we should adjust the plugin to use the same method:

```
// Get the site domain and get rid of www.
$sitename = wp_parse_url( network_home_url(), PHP_URL_HOST );
if ( 'www.' === substr( $sitename, 0, 4 ) ) {
	$sitename = substr( $sitename, 4 );
}

$from_email = 'wordpress@' . $sitename;
```

**Comment by szepeviktor** [comment]
> I guess that means we should adjust the plugin to use the same method:

Yes.

**Comment by benhuson** [comment]
I think I'll need to make sure the updated way of getting the site name is only used in WordPress version from when this was changed in core WP.

**Comment by benhuson** [comment]
Fixed in https://github.com/benhuson/wp-mailfrom/commit/811537d5ba982daad6821fc16a2f73371f410c03

**Comment by szepeviktor** [comment]
Thank you.
